enhancements/baremetal/minimise-baremetal-footprint.md

```diff
@@ -0,0 +1,143 @@
+see-also: compact-clusters
+replaces:
+superseded-by:
+---
+
+# Minimise Baremetal footprint
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Over recent releases OpenShift has improved support for small-footprint
+deployments, in particular with the compact-clusters enhancement which adds
+full support for 3-node clusters where the masters are schedulable.
+
+This is a particularly useful deployment option for baremetal PoC environments,
+where often the amount of physical hardware is limited, but there is still the
+problem of where to run the installer/bootstrap-VM in this environment.
+
+The current solution for IPI baremetal is to require a 4th bootstrap host,
+which is a machine physically connected to the 3 master nodes, that runs
+the installer and/or the bootstrap VM.  This effectively means the minimum
+footprint is 4 nodes, unless you can temporarily connect a provisioning host
+to the cluster machines.
+
+A similar constraint exists for UPI baremetal deployments, where although a
+3 master cluster is possible, you need to run a 4th bootstrap node somewhere
+for the duration of the initial installation.
+
+## Motivation
+
+This proposal outlines a potential approach to avoid the requirement for a
+4th node, leveraging the recent etcd-operator improvements and work to enable
+a live-iso replacement for the bootstrap VM.
+
+### Goals
+
+* Enable compact clusters to be deployed on baremetal with exactly 3 nodes
+* Avoid the need for additional nodes to run install/bootstrap components
+
+### Non-Goals
+
+* Supporting any topology other than three masters for production clusters.
+* Dupporting deployment of a single master or scaling from such a deployment.
+
+## Proposal
+
+### User Stories
+
+As a user of OpenShift, I should be able to install a 3-node cluster (no workers)
+in baremetal environments, without the requirement to temporarily connect a 4th
+node to host installer/bootstrap services.  Instead I want to boot one of the
+3 target nodes with an image which enables installation to proceed, and the
+end state should be a full supportable 3-node cluster.
+
+
+### Risks and Mitigations
+
+This proposal builds on work already completed e.g etcd-operator improvements
+but we need to ensure any change in deployment topology is well tested and
+fully supported, to avoid compact-baremetal deployments being an unreliable
+corner-case.
+
+## Design Details
+
+### Enabling three-node clusters on baremetal
+
+OpenShift now provides a bootable RHCOS based installer ISO image, which can
+be booted on baremetal, and adapted to install the components normally
+deployed on the bootstrap VM [TODO reference to PoC @avishayt]
```

Comment:
HLD doc: https://docs.google.com/document/d/1jxNMTlotmJ0GFZ1GUEQ3hfOOVLzMT6mvY8Ufo5RdErY/edit?usp=sharing

Reply:
I think we need to refactor the design from there (parts not covered in this enhancement) into one or more additional enhancements?

Reply:
This is now in-progress via https://github.com/openshift/enhancements/pull/376